source/common/common/cleanup.h

```diff
@@ -14,4 +15,18 @@ class Cleanup {
   std::function<void()> f_;
 };
 
+// RAII helper class to add an element to an std::list on construction and erase
+// it on destruction.
+template <class T> class ListAddAndRemove {
+public:
+  ListAddAndRemove(std::list<T>& container, T element) : container_(container) {
+    it_ = container.emplace(container.begin(), element);
```

Comment:
Are there any operations on `container` that would invalidate the `it_` reference? If so, maybe document them in the constructor docs, since these are useful to know.

Reply:
It can be invalidated by clearing the list or deleting the element some other way. I added a comment and also methods to allow explicitly deleting the element before the destructor, or canceling the delete in case the iterator was invalidated. This allowed be to convert over another use case (GrpcMuxWatchImpl).